### ISSUE TITLE
Sequence contains no matching element

### ISSUE THREAD

**Issue by LaurenceMillward** [mRemoteNG/mRemoteNG#483]
Heya,

Trying to use the SQL feature so I installed SQL Server 2016 on Windows Server 2016, created the tables as per the SQL script, entered the SQL details into mRemoteNG and it seemed good, 5 seconds later the app crashed "MRemoteNG has stopped working". Upon re-opening the app gives an SQL error saying "Sequence contains no matching element "

![sqlmremote](https://cloud.githubusercontent.com/assets/26876521/24617366/95cc68da-188b-11e7-89da-63b64ac2679c.PNG)

Seen this before or know of a way to fix? I've deleted all the tables and re-created them but it's not liking anything to do with SQL. 

Using the latest development version (downloaded and installed today).

**Comment by sparerd** [comment]
Just to confirm, can you please provide the exact version number of the build you are running?

**Comment by LaurenceMillward** [comment]
Hi,

v1.75.7003.25190 is the current version.

I just installed it to another PC running windows 10 and have the same issue.

**Comment by sparerd** [comment]
Thanks for confirming - 

Run the following 2 queries on your sql instance and paste the output here:
```sql
select ORDINAL_POSITION,COLUMN_NAME,DATA_TYPE
from INFORMATION_SCHEMA.COLUMNS
where TABLE_NAME = 'tblCons'
```

```sql
select ConfVersion
from tblRoot
```

**Comment by LaurenceMillward** [comment]
No problem; hopefully GitHub doesn't break the formatting, here is the output: 

ORDINAL_POSITION	COLUMN_NAME	DATA_TYPE
1	ID	int
2	ConstantID	varchar
3	PositionID	int
4	ParentID	varchar
5	LastChange	datetime
6	Name	varchar
7	Type	varchar
8	Expanded	bit
9	Description	varchar
10	Icon	varchar
11	Panel	varchar
12	Username	varchar
13	DomainName	varchar
14	Password	varchar
15	Hostname	varchar
16	Protocol	varchar
17	PuttySession	varchar
18	Port	int
19	ConnectToConsole	bit
20	UseCredSsp	bit
21	RenderingEngine	varchar
22	ICAEncryptionStrength	varchar
23	RDPAuthenticationLevel	varchar
24	RDPMinutesToIdleTimeout	int
25	RDPAlertIdleTimeout	bit
26	Colors	varchar
27	Resolution	varchar
28	DisplayWallpaper	bit
29	DisplayThemes	bit
30	EnableFontSmoothing	bit
31	EnableDesktopComposition	bit
32	CacheBitmaps	bit
33	RedirectDiskDrives	bit
34	RedirectPorts	bit
35	RedirectPrinters	bit
36	RedirectSmartCards	bit
37	RedirectSound	varchar
38	SoundQuality	varchar
39	RedirectKeys	bit
40	Connected	bit
41	PreExtApp	varchar
42	PostExtApp	varchar
43	MacAddress	varchar
44	UserField	varchar
45	ExtApp	varchar
46	VNCCompression	varchar
47	VNCEncoding	varchar
48	VNCAuthMode	varchar
49	VNCProxyType	varchar
50	VNCProxyIP	varchar
51	VNCProxyPort	int
52	VNCProxyUsername	varchar
53	VNCProxyPassword	varchar
54	VNCColors	varchar
55	VNCSmartSizeMode	varchar
56	VNCViewOnly	bit
57	RDGatewayUsageMethod	varchar
58	RDGatewayHostname	varchar
59	RDGatewayUseConnectionCredentials	varchar
60	RDGatewayUsername	varchar
61	RDGatewayPassword	varchar
62	RDGatewayDomain	varchar
63	InheritCacheBitmaps	bit
64	InheritColors	bit
65	InheritDescription	bit
66	InheritDisplayThemes	bit
67	InheritDisplayWallpaper	bit
68	InheritEnableFontSmoothing	bit
69	InheritEnableDesktopComposition	bit
70	InheritDomain	bit
71	InheritIcon	bit
72	InheritPanel	bit
73	InheritPassword	bit
74	InheritPort	bit
75	InheritProtocol	bit
76	InheritPuttySession	bit
77	InheritRedirectDiskDrives	bit
78	InheritRedirectKeys	bit
79	InheritRedirectPorts	bit
80	InheritRedirectPrinters	bit
81	InheritRedirectSmartCards	bit
82	InheritRedirectSound	bit
83	InheritSoundQuality	bit
84	InheritResolution	bit
85	InheritUseConsoleSession	bit
86	InheritUseCredSsp	bit
87	InheritRenderingEngine	bit
88	InheritICAEncryptionStrength	bit
89	InheritRDPAuthenticationLevel	bit
90	InheritRDPMinutesToIdleTimeout	bit
91	InheritRDPAlertIdleTimeout	bit
92	InheritUsername	bit
93	InheritPreExtApp	bit
94	InheritPostExtApp	bit
95	InheritMacAddress	bit
96	InheritUserField	bit
97	InheritExtApp	bit
98	InheritVNCCompression	bit
99	InheritVNCEncoding	bit
100	InheritVNCAuthMode	bit
101	InheritVNCProxyType	bit
102	InheritVNCProxyIP	bit
103	InheritVNCProxyPort	bit
104	InheritVNCProxyUsername	bit
105	InheritVNCProxyPassword	bit
106	InheritVNCColors	bit
107	InheritVNCSmartSizeMode	bit
108	InheritVNCViewOnly	bit
109	InheritRDGatewayUsageMethod	bit
110	InheritRDGatewayHostname	bit
111	InheritRDGatewayUseConnectionCredentials	bit
112	InheritRDGatewayUsername	bit
113	InheritRDGatewayPassword	bit
114	InheritRDGatewayDomain	bit
115	LoadBalanceInfo	varchar
116	AutomaticResize	bit
117	InheritLoadBalanceInfo	bit
118	InheritAutomaticResize	bit


ConfVersion
2.6

**Comment by sparerd** [comment]
Thanks - that data looks as it should for this version. I'm not able to reproduce the error on sql server 2012. I'll get 2016 set up and give that a go. In the meantime, could you also provide the mremoteng.log file?

Log file details [can be found here](https://github.com/mRemoteNG/mRemoteNG/wiki/Troubleshooting-and-Logging) if you need.

**Comment by LaurenceMillward** [comment]
Heya,

Here is the log file. Not sure if it provides anything useful for you, i've closed and opened the application a few times to try and see if something pops up (other than the Sequence contains no matching element)

```
2017-04-04 16:22:56,350 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:22:56,362 [1] INFO - Command Line: System.String[]
2017-04-04 16:22:57,690 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:22:57,690 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:22:57,691 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:22:57,693 [1] INFO - Checking FIPS Policy...
2017-04-04 16:22:57,694 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:22:58,602 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:22:58,602 [1] INFO - Adding External App: Password cmd /k echo %password%
2017-04-04 16:22:59,090 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-04 16:23:31,815 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:23:31,826 [1] INFO - Command Line: System.String[]
2017-04-04 16:23:32,926 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:23:32,926 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:23:32,926 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:23:32,927 [1] INFO - Checking FIPS Policy...
2017-04-04 16:23:32,928 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:23:34,353 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:23:34,354 [1] INFO - Adding External App: Password cmd /k echo %password%
2017-04-04 16:23:34,858 [1] ERROR- Load from SQL failed
Sequence contains no matching element
2017-04-04 16:23:36,938 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-04 16:24:04,784 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:24:04,793 [1] INFO - Command Line: System.String[]
2017-04-04 16:24:05,888 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:24:05,889 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:24:05,889 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:24:05,889 [1] INFO - Checking FIPS Policy...
2017-04-04 16:24:05,890 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:24:07,263 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:24:07,264 [1] INFO - Adding External App: Password cmd /k echo %password%
2017-04-04 16:24:07,754 [1] ERROR- Load from SQL failed
Sequence contains no matching element
2017-04-04 16:24:09,911 [1] ERROR- Load from SQL failed
Sequence contains no matching element
2017-04-04 16:24:13,075 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-04 16:25:37,507 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:25:37,518 [1] INFO - Command Line: System.String[]
2017-04-04 16:25:38,821 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:25:38,821 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:25:38,822 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:25:38,822 [1] INFO - Checking FIPS Policy...
2017-04-04 16:25:38,824 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:25:39,641 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:25:39,641 [1] INFO - Adding External App: Password cmd /k echo %password%
2017-04-04 16:25:40,060 [1] ERROR- Load from SQL failed
Sequence contains no matching element
2017-04-04 16:25:45,916 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-04 16:26:00,388 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:26:00,397 [1] INFO - Command Line: System.String[]
2017-04-04 16:26:01,510 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:26:01,510 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:26:01,511 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:26:01,512 [1] INFO - Checking FIPS Policy...
2017-04-04 16:26:01,514 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:26:02,896 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:26:03,472 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-04 16:26:21,802 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-04 16:26:21,812 [1] INFO - Command Line: System.String[]
2017-04-04 16:26:22,910 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-04 16:26:22,910 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-04 16:26:22,910 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-04 16:26:22,911 [1] INFO - Checking FIPS Policy...
2017-04-04 16:26:22,912 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-04 16:26:24,292 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-04 16:26:24,791 [1] ERROR- Load from SQL failed
Sequence contains no matching element


```

**Comment by LaurenceMillward** [comment]
C:\Users\All Users\Microsoft\Windows\WER\ReportArchive attached



[ReportArchive.zip](https://github.com/mRemoteNG/mRemoteNG/files/893681/ReportArchive.zip)

**Comment by sparerd** [comment]
Thanks for that - it helps narrow down a little bit. This is likely a LINQ query that is failing

**Comment by sparerd** [comment]
So the error is likely being thrown from one of several places:
- DataTableSerializer.Serialize() : line 28
- DataTableDeserializer.CreateNodeHierarchy() : line 183, 188

I'll try to recreate the issue as that will give the best chance of finding the root cause.

**Comment by LaurenceMillward** [comment]
Happy to help; if you need anything else from me let me know.

**Comment by sparerd** [comment]
I still wasn't able to recreate the issue on a live system, but I added some more safety checking in the sections that I think are causing the problem. There were a few "brittle" sections that should handle unexpected situations better now.

If you are able to build from source, the new changes are available on the branch `483_make_datatable_serializer_safer`. Otherwise, I'm hoping we will have a new minor/hotfix release out this next week.

**Comment by wrufeger** [comment]
Same issue here after upgrading table shema from 2.5 to 2.6

**Comment by LaurenceMillward** [comment]
Hi,

Just downloaded the hotfix 4 release and still having the same issue as before. Soon as I tick the box to use SQL it crashes the program, upon re-opening the program I get the "Sequence contains no matching element". 

I have tested the SQL connection and can connect fine through other programs. Tried with the SA account within SQL too just incase it was a user permissions error. 

`2017-04-26 15:13:21,510 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-26 15:13:21,518 [1] INFO - Command Line: System.String[]
2017-04-26 15:13:22,796 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-26 15:13:22,796 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-26 15:13:22,797 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-26 15:13:22,800 [1] INFO - Checking FIPS Policy...
2017-04-26 15:13:22,802 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-26 15:13:23,582 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-26 15:13:24,016 [1] WARN - XmingPortablePuttySessions.Watcher.StartWatching() failed.
The directory name sessions is invalid.
2017-04-26 15:13:51,263 [1] INFO - mRemoteNG 1.75.7003.25190 starting.
2017-04-26 15:13:51,273 [1] INFO - Command Line: System.String[]
2017-04-26 15:13:52,381 [1] INFO - Microsoft Windows 10 Home  64-bit
2017-04-26 15:13:52,382 [1] INFO - Microsoft .NET CLR 4.0.30319.42000
2017-04-26 15:13:52,382 [1] INFO - System Culture: en-GB/English (United Kingdom)
2017-04-26 15:13:52,384 [1] INFO - Checking FIPS Policy...
2017-04-26 15:13:52,386 [1] INFO - Checking Lenovo AutoScroll Utility...
2017-04-26 15:13:53,711 [1] INFO - Loading External Apps from: C:\Users\laure\AppData\Roaming\mRemoteNG\extApps.xml
2017-04-26 15:13:55,222 [1] ERROR- Load from SQL failed
Sequence contains no matching element`
